### PR TITLE
feat: トークナイザ

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
# GitHub Review — BERT日本語トークナイズ（#4-4〜#4-6）
対象コード：
- 4-4: tokenizer.tokenize('明日は自然言語処理の勉強をしよう。')
- 4-5: tokenizer.tokenize('明日はマシンラーニングの勉強をしよう。')
- 4-6: tokenizer.tokenize('機械学習を中国語にすると机器学习だ。')

## 概要（総評）
本コードは `BertJapaneseTokenizer` の挙動を観察する最小例として適切です。評価観点としては、（1）形態素解析→WordPieceの二段階分割の理論背景の明記、（2）Whole Word Masking（WWM）と推論時トークナイズの関係の切り分け、（3）辞書・語彙・Unicode正規化がトークン列に与える影響の再現性確保、（4）下流タスク（分類/トークン分類/QA）へ接続するための「tokenizeではなくencode（`return_tensors='pt'`）」の推奨、が主です。以下に理論中心で指摘と改善提案をまとめます。

---

## 理論的な主要論点

### 1) 日本語BERTにおけるトークナイズの二段構え
- **形態素解析（語単位推定）→WordPiece分割（語彙に基づくサブワード化）**の二段階で行われます。  
- 形態素解析は辞書（ipadic / unidic系）やそのバージョンにより結果が変動します。  
- WordPieceは語彙外（OOV）に対して細分化または `[UNK]` を返します。CJKの広いカバレッジにより、日本語＋中国語混在でも多くは細分化で拾えますが、辞書・語彙に依存します。  
**実務的含意**：トークン列の再現性は「トークナイザ実装＋辞書種類とバージョン＋transformersバージョン」に依存するため、ロックと記録が必要です。

### 2) Whole Word Masking（WWM）の位置づけ
- **WWMは事前学習時のマスク戦略**であり、**推論時の `tokenize` 結果は直接変わりません**。  
- 効果は「同一の“語”（形態素のまとまり）に属するサブワードを一括でマスク」→語レベルの完形復元を強制し、より語彙レベルの意味一貫性を得やすい点にあります。  
- ただしWWMの有無は**学習済み表現**（パラメータ）に反映されるため、**同じ入力でもモデルの応答・性能**は影響を受けます。  
**実務的含意**：推論側では`tokenize`の表示より、**対応するWWM版のモデル重み**を使うことが本質です（tokenizerとのペア整合性が重要）。

### 3) 文種別の着眼点（今回3例）
- **4-4**「自然言語処理」は複合名詞。形態素解析の辞書次第で「自然/言語/処理」「自然言語/処理」など分割差が出ます。その後のWordPieceで更に細分化されうるため、**分割は固定的でない**点に注意。  
- **4-5**「マシンラーニング」はカタカナ複合語。**カタカナ語は単語/サブワードの切れ目が辞書と語彙次第**になりやすい。  
- **4-6** 末尾「机器学习」は中国語（簡体字）。**日本語辞書では未知語扱い**になりがちで、**字単位サブワード**への分解や稀に`[UNK]`が起こりえます。CJK混在時は**Unicode正規化や全角/半角差**の影響も検討対象です。

### 4) 下流タスクの観点：`tokenize`より`__call__`（encode）を使う理由
- 実運用では `tokenizer(text, return_tensors='pt', padding=..., truncation=...)` を用い、  
  **`input_ids` / `attention_mask` / `token_type_ids`** を**テンソル**で得るのが標準です。  
- **トークン分類（NER等）**では `is_split_into_words=True` と `word_ids()` を使って、**語→サブワード**のアライメントを安定的に扱います。  
- **文分類**や**QA**でも特殊トークン `[CLS]` / `[SEP]` の位置を意識したテンソル処理が必要で、`tokenize`単体の観察からは漏れがちです。  
**実務的含意**：**`tokenize`は可視化・デバッグ向け**。**学習・推論コードでは`return_tensors='pt'`のパス**に統一した方が安全で、下流処理と整合します。

### 5) 再現性・評価設計
- **辞書**（ipadic / unidic-lite）と**transformers**の**バージョン固定**、`normalization`のルール明記（例：NFKCで幅記号正規化）を推奨します。  
- CIで**不変量（invariants）テスト**を推奨：  
  - 例）`len(convert_ids_to_tokens(convert_tokens_to_ids(tokens))) == len(tokens)` （ただしOOVでの融合/分裂に注意）  
  - 例）`attention_mask.sum()` が非PAD長と一致  
  - 例）`token_type_ids` が単文で全ゼロ  
- **ベンチ**：長文やCJK混在、記号類（全角/半角/句読点）の差分ケースを網羅したサンプルセットで**回帰テスト**を持つと安全です。

---

## アクション可能な提案（理論に基づく改善）

1. **学習/推論の標準入出力へ寄せる**  
   - `tokenize`呼び出しは**可視化用途**に限定し、**実務サンプル**としては `tokenizer(..., return_tensors='pt')` を提示してください。  
   - 下流タスク（分類/トークン分類/QA）に応じたテンソルの使い方（`[CLS]`の抽出、`word_ids()`でのラベル拡張など）を**コメントで理論と紐づけ**て明記すると有用です。

2. **辞書・正規化・バージョンの固定と記録**  
   - `requirements.txt` 例：`transformers==4.xx.x`, `fugashi==x.x.x`, `ipadic==x.x.x` など。  
   - 入力前に `unicodedata.normalize("NFKC", text)` を（必要なら）明示。**カタカナ・記号・全半角**の影響を減らします。

3. **CJK混在（4-6）の評価基準を言語横断で設計**  
   - 「中国語部分が `[UNK]` にならないこと」ではなく、**安定に下流へ影響しない形**（例：字単位サブワードで一定長以下、など）の**不変量**を定めると堅牢です。  
   - もし中国語の語彙カバレッジが重要なら、日本語BERTではなく**多言語モデル（mBERT, XLM-R）**の比較も検討を示すと親切です。

4. **テストの目的を「分割の完全一致」から「性質の一致」へ**  
   - 形態素解析の辞書更新で**分割境界は将来的に変わりうる**ため、**完全一致テスト**は脆いです。  
   - 代わりに、**最大長を超えない**、**PADが正しくマスクされる**、**特殊トークンの配置が正しい**、**（トークン分類時）ラベル展開が整合**などの**性質テスト**を増やしてください。

5. **コメントに理論式を最小限添える**  
   - Self-Attention の要点：`softmax(QK^T / sqrt(d_k)) V`  
   - WWMの要点：**同一語に属するサブワード全体を一括マスク**（＝語レベルの完形復元圧力）。  
   - こうした式と一文の結びつけで「なぜこの前処理が必要か」を読者が再確認できます。

---

## 具体的な改善案（サンプル差し替え例）
下記は**観察（tokenize）**を残しつつ、**実運用（テンソル化）**も併記する例です。理論との結合点をコメントで補足しています。

# --- 修正案サンプル（置換用） ---
from transformers import BertJapaneseTokenizer
import unicodedata

# 依存：transformers / fugashi / ipadic（固定推奨）
MODEL = "tohoku-nlp/bert-base-japanese-whole-word-masking"
tokenizer = BertJapaneseTokenizer.from_pretrained(MODEL)

def normalize(text: str) -> str:
    # 理由：CJK混在や全角/半角差の影響を抑える（再現性と頑健性）
    return unicodedata.normalize("NFKC", text)

texts = [
    "明日は自然言語処理の勉強をしよう。",
    "明日はマシンラーニングの勉強をしよう。",
    "機械学習を中国語にすると机器学习だ。",
]

# A) 可視化（研究・デバッグ用）：分割の様子を観察
for t in texts:
    toks = tokenizer.tokenize(normalize(t))
    # 形態素→WordPieceの二段分割。分割は辞書と語彙に依存し再現性に注意。
    # WWMは事前学習の戦略であり、ここでの分割自体は直接変わらない。
    # print(toks)

# B) 実運用（学習・推論用）：テンソル化してモデルへ渡す前処理
batch = tokenizer(
    [normalize(t) for t in texts],
    return_tensors="pt",      # ← 下流タスクへ直結できるテンソル出力
    padding=True,             # ← バッチ整列。PADはattention_maskで無視
    truncation=True,          # ← 最大長超過時の安全弁
    max_length=64,
)
# 不変量の例：
# - 単文なので token_type_ids は原則ゼロ列
# - attention_mask.sum(dim=1) が各文の有効トークン長と一致
# - convert_ids_to_tokens で可逆性（おおむね）を確認（OOV注意）
# print(batch["input_ids"].shape, batch["attention_mask"].shape, batch["token_type_ids"].unique())

---

## リスク・注意点
- **形態素解析辞書の差分**で分割が変わるため、期待トークン列の完全一致を要件にすると運用が不安定になります。  
- **CJK混在**では言語境界を跨ぐため、**日本語BERTの語彙でどこまで拾えるか**を前提設計に反映させてください。多言語モデル比較の注記が望ましいです。  
- **WWMの勘違い**に注意：推論のトークナイズ自体は変わらず、学習済み表現に効く点を明示してください。

---

## マイナー（記述の小改善）
- 変数名を `tokens_44/45/46` として保持するのはOKですが、未使用だと静的解析で警告の可能性があるため、デバッグプリントや簡易検証（長さ・`[UNK]`有無）を付けるとよいです。  
- コメントに「形態素解析→WordPiece」と「WWMは学習側の戦略」の二点を太字（または強調）で固定記述すると、後から読んだ人にも明瞭です。  
- READMEやdocstringに**要再現設定（依存バージョン/辞書/正規化）**をまとめると、将来の差分解析が楽になります。

---

## 結論
本サンプルは「BertJapaneseTokenizerの分割を目視する」という目的に合致しています。理論中心に最小追加を行うなら、（1）WWMは推論トークナイズに直結しない旨の明記、（2）辞書・バージョン・正規化で再現性が揺らぐ点の明記、（3）下流タスク用の`return_tensors='pt'`経路の提示、の3点を加えると、研究コードとしても実務コードとしても説得力が増します。
